src/ssr-template-helper.js

```diff
@@ -58,29 +67,42 @@ export class SsrTemplateHelper {
   }
 
   /**
-   * Proxies xhr and template rendering to the viewer and renders
-   * the response.
+   * Proxies xhr and template rendering to the viewer and renders the response.
    * @param {!Element} element
+   * @param {!./service/xhr-impl.FetchData} fetchData The fetch/XHR related data.
+   * @param {?SsrTemplateDef=} opt_templates Response templates to pass into
+   *     the payload. If provided, finding the template in the passed in
+   *     element is not attempted.
    * return {!Promise<{data:{?JsonObject|string|undefined}}>}
    */
-  fetchAndRenderTemplate(element) {
+  fetchAndRenderTemplate(element, fetchData, opt_templates = null) {
     const inputsAsJson = this.getElementInputsAsJson_(element);
     const elementAttrsAsJson =
         this.getElementAttributesAsJson_(element);
     elementAttrsAsJson['inputData'] = inputsAsJson;
-    const template = this.templates_.maybeFindTemplate(element);
     let mustacheTemplate;
-    if (template) {
-      // The document fragment can't be used in the message channel API thus
-      // serializeToString for a string representation of the dom tree.
-      mustacheTemplate = this.xmls_.serializeToString(
-          this.templates_.findTemplate(element));
+    if (!opt_templates) {
+      const template = this.templates_.maybeFindTemplate(element);
+      if (template) {
+        // The document fragment can't be used in the message channel API thus
+        // serializeToString for a string representation of the dom tree.
+        mustacheTemplate = this.xmls_.serializeToString(
+            this.templates_.findTemplate(element));
+      }
     }
     const data = dict({
+      'originalRequest':
+          toStructuredCloneable(fetchData.xhrUrl, fetchData.fetchOpt),
       'data': elementAttrsAsJson,
```

Comment:
do we still need data though? the reason i had data originally was because I was under the impression that you would construct the original request on your end, which was wrong, so this shouldn't be needed. Pls confirm.

Reply:
Ideally I would like the structure to be:
{
  'originalRequest':
  'ampComponent': {
    'type': amp-list or amp-form,
    'successTemplate': {
      'type': //only amp-mustache for now
      'payload':
    }
    'errorTemplate': {
      'type':
      'payload':
    }
    'ampListAttribute': {
      'items':
      'singleItem':
      'maxItems':
    }
  }
}
This would be the easiest for the server. But any similar one would be fine as we are doing a conversion on the server anyway.

Same for the response:
{
  'orginalResponse' or 'httpResponse': //this is tricky as it's not compatible with xhr-impl now, we may need to discuss more.
  'ampTemplateOutput': {
    'html':
  }
}

Let me know if this makes sense to you! Happy to chat about it!